docs/labs/redos.html

```diff
@@ -0,0 +1,159 @@
+- present: |
+    matches \( /.*[^$]/
+  text: Match the whole string - end the regular expression with $
+- present: |
+    matches \( /\^\[A-Z\]
+  text: That would match only letters, you need digits as well.
+- present: |
+    matches \( /\^\[a-z\]
+  text: That would match only lower case letters, the format requirement is uppercase letters.
+- present: |
+    matches \( /\^\(\[A-Z0-9\]\+\)\+\$ 
+  text: Remember to fix the regex, the outer + quantifier causes backtracking by trying to match one or more sequences of one or more uppercase alphanumeric characters.
+- present: |
+    matches \( /\^\(\[A-Z0-9\]\+\)\$ 
+  text: Rremove the grouping, you don’t need the parentheses.
+
+# debug: true
+</script>
+</head>
+<body>
+<!-- For GitHub Pages formatting: -->
+<div class="container-lg px-3 my-5 markdown-body">
+<h1>Lab Exercise ReDoS</h1>
+<p>
+This is a lab exercise on developing secure software.
+For more information, see the <a href="introduction.html" target="_blank">introduction to
+the labs</a>.
+
+<p>
+<h2>Task</h2>
+<p>
+Learn how to identify and fix a code susceptible to ReDoS attack 
+
+<p>
+<h2>Background</h2>
+<p>
+Regular expressions are a way to do input validation, checking whether data matches a specific pattern. Poorly designed expressions may result in potential denial of service conditions (ReDoS ). This attack exploits the fact that most Regular Expression implementations may reach extreme situations, so an attacker can force situations where the regular expression can be run for an extremely long time, exponentially based on the input size.
+<p>
+In this exercise we are going to fix an input validation that uses regular expressions but is susceptible to ReDos attack.
+
+
+<p>
+<h2>Task Information</h2>
+<p>
+
+<p>
+The code below sets up handlers for a get request on path <tt>/parts</tt>. This code could be triggered, for example, by requesting <tt>http://localhost:3000/part?id=AB123</tt> (if it was running at localhost and responding to port 3000). If there are no validation errors, the code is supposed to show the part id. If there is a validation error, it responds with HTTP error code 422 ("Unprocessable Content"), a status code suggesting that the request was invalid for some reason, along with an error message. 
+<p>
+For this lab we want to fix a regex that is susceptible to a ReDos attack, applying these countermeasures: 
+<ol>
+<li>Limit the maximum length of input strings and check the input length first. 
+  <ol>
+    <li>After selecting the id parameter (<tt>query('id')</tt>), use the validation requirement <tt>isLength()</tt> to limit the string size. We'll need to use the maximum optional parameter to set the maximum size required: <tt>isLength({max: YOUR_MAXIMUM})</tt></li>
+  </ol>
+</li>
+<li>Modify the regex so that it doesn’t have the worst-case behavior. Be especially wary of any group “(...)” that contains a branch and/or ends with a repeat and is itself repeated</li>
```

Comment:
Add a final period (.), or remove it from previous item (`first.`) for consistency.